pkg/trace/obfuscate/sql.go

```diff
@@ -342,3 +344,73 @@ func (o *Obfuscator) obfuscateSQL(span *pb.Span) {
 	}
 	traceutil.SetMeta(span, sqlQueryTag, oq.Query)
 }
+
+type mapEntryTransform func(key string, value interface{}) (string, interface{})
+
+// transformMapEntriesRecursive returns a copy of the map object, recursively transforming any keys matching keyPattern
+func transformMapEntriesRecursive(obj interface{}, keyPattern *regexp.Regexp, transform mapEntryTransform) interface{} {
+	if mapObj, ok := obj.(map[string]interface{}); ok {
+		result := make(map[string]interface{})
+		for key, val := range mapObj {
+			if keyPattern.MatchString(key) {
+				key, val = transform(key, val)
+				// transform can decide to drop a key by returning empty string for key
+				if key != "" {
+					result[key] = val
+				}
+			} else {
+				result[key] = transformMapEntriesRecursive(val, keyPattern, transform)
+			}
+		}
+		return result
+	} else if listObj, ok := obj.([]interface{}); ok {
+		result := make([]interface{}, len(listObj), len(listObj))
+		for i, val := range listObj {
+			result[i] = transformMapEntriesRecursive(val, keyPattern, transform)
+		}
+		return result
+	}
+	return obj
+}
+
+var planObfuscationPattern = regexp.MustCompile(`attached_condition|(\w+ Cond)`)
```

Comment:
Regex feels like the wrong solution here and prone to unexpected behavior over time, especially as we add more dbms. Is there any reason not to enumerate the list of keys in a map? I think that would make the code easier to read/modify and less prone to bugs.

- `attached_condition`
- `Index Cond`
- `Recheck Cond`
- ...

Reply:
Good idea! Updated to use an explicit list of keys for both obfuscation and normalization.